compiler/rustc_mir_transform/src/check_unsafety.rs

```diff
@@ -100,13 +100,15 @@ impl<'tcx> Visitor<'tcx> for UnsafetyChecker<'_, 'tcx> {
             | StatementKind::StorageLive(..)
             | StatementKind::StorageDead(..)
             | StatementKind::Retag { .. }
-            | StatementKind::AscribeUserType(..)
             | StatementKind::PlaceMention(..)
             | StatementKind::Coverage(..)
             | StatementKind::Intrinsic(..)
             | StatementKind::Nop => {
                 // safe (at least as emitted during MIR construction)
             }
+            // `AscribeUserType` just exists to help MIR borrowck.  It has no semantics, and
+            // everything is already reported by `PlaceMention`.
+            StatementKind::AscribeUserType(..) => return,
```

Comment:
FWIW this contradicts what you are doing in the other PR where this statement becomes relevant for Miri. If Miri checks it, it should probably also be unsafe. If this is truly only a type system guide without a possibility of introducing UB, Miri should ignore it.